core/src/test/java/com/google/errorprone/bugpatterns/BadImportTest.java

```diff
@@ -388,4 +388,63 @@ public void doesNotMatchProtos() {
             "}")
         .doTest();
   }
+
+  @Test
+  public void badEnclosingTypes() {
+    refactoringTestHelper
+        .setArgs("-XepOpt:BadImport:BadEnclosingTypes=org.immutables.value.Value")
+        .addInputLines(
+            "org/immutables/value/Value.java",
+            "package org.immutables.value;",
+            "",
+            "public @interface Value {",
+            "  @interface Immutable {}",
+            "}")
+        .expectUnchanged()
+        .addInputLines(
+            "Test.java",
+            "import org.immutables.value.Value.Immutable;",
+            "",
+            "@Immutable",
+            "interface Test {}")
+        .addOutputLines(
+            "Test.java",
+            "import org.immutables.value.Value;",
+            "",
+            "@Value.Immutable",
+            "interface Test {}")
+        .doTest();
+  }
+
+  @Test
+  public void badEnclosingTypes_doesNotMatchFullyQualifiedName() {
+    compilationTestHelper
+        .setArgs("-XepOpt:BadImport:BadEnclosingTypes=org.immutables.value.Value")
+        .addSourceLines(
+            "org/immutables/value/Value.java",
+            "package org.immutables.value;",
+            "",
+            "public @interface Value {",
+            "  @interface Immutable {}",
+            "}")
+        .addSourceLines("Test.java", "@org.immutables.value.Value.Immutable", "interface Test {}")
+        .doTest();
+  }
+
+  @Test
+  public void badEnclosingTypes_staticMethod() {
+    compilationTestHelper
+        .setArgs("-XepOpt:BadImport:BadEnclosingTypes=com.google.common.collect.ImmutableList")
+        .addSourceLines(
+            "Test.java",
+            "import static com.google.common.collect.ImmutableList.toImmutableList;",
+            "import com.google.common.collect.ImmutableList;",
+            "import java.util.stream.Collector;",
+            "",
+            "class Test {",
+            "  // BUG: Diagnostic contains: ImmutableList.toImmutableList()",
+            "  Collector<?, ?, ImmutableList<Object>> immutableList = toImmutableList();",
```

Comment:
Is disallowing static import desirable? Should we change the behavior only for NestedTypes? :thinking:  https://github.com/google/error-prone/blob/63cf19274e382089693102a19bf31d7dbe791807/core/src/main/java/com/google/errorprone/bugpatterns/BadImport.java#L107-L109

Reply:
The current behaviour of treating static imports consistently seems OK to me. Obviously `toImmutableList` is just an example, in general if `com.example.Foo` is supposed to provide a namespace and not have its members be imports, it seems OK to discourage both `import com.example.Foo.Bar` and `import static com.example.Foo.baz`.